T 09 item backend

Ali-Alkhars opened the pull request:
Hello everyone! T-09 (implementing the add, delete and edit buttons for choice board items) is done.

### These are the main things I added:
- add item widget
- edit item widget
- delete item alert dialogue
- A class for firebase manipulation functions (handling backend)
- Widget tests and Unit tests for all added classes
- Had to do lots of refactoring to make classes able to take mock firebase functions for testing

_**Note:**_ Manual testing is very tricky because loading data from the database is not yet implemented. If you would like to manually test:
- go to our firestore console 
- manually add three categories (with correct fields)
- Take the ids of these created categories
- Go to flutter: lib/data/choice_boards_data.dart
- Paste one id to each Draggable list (change the current data)
- Create items using the app (add an item) buttons
- Copy the ids of the items from the firestore console
- Paste them into the ids of the Draggable list items
- Now you should also be able to edit/delete items using the app buttons

Ali-Alkhars committed:
make add item button open a popup to take user input

- Add a widget to create a custom popup hero
- Create a page for add item popup
- Install the popup card library

Ali-Alkhars committed:
take user input for creating a new item

Ali-Alkhars committed:
implement adding items to firestore database

Ali-Alkhars committed:
Add ability to upload/take images

- extended the add choice board item page
- Added image_picker and path_provider libraries
- Added instructions to ask for permissions for uploading/taking images

Ali-Alkhars committed:
Save images to cloud storage and save url in firestore

- Add comments
- Add the firebase_storage library

Ali-Alkhars committed:
temprary createItem implementation

Ali-Alkhars committed:
Fix error handling for creating items

Ali-Alkhars committed:
Fix items permissions error

Ali-Alkhars committed:
Fix permission denied error for categoryItems

Ali-Alkhars committed:
Switch delete and edit buttons for categories

Ali-Alkhars committed:
Connect add item buttons to categories

- Edit the draggable_list model to contain item id and category id
- Create a custom widget in the widgets folder to represent the add item button
- Delete the add item method in admin_choice_boards.dart and use the new widget instead

Ali-Alkhars committed:
Make creating a new categoryItem take an appropriate rank

Ali-Alkhars committed:
Refactor pick image buttons into a separate widget in the widgets folder

Ali-Alkhars committed:
Implement deleting an item

- Create a separate widget in the widgets folder "delete_item_button.dart"
- Update the admin choice boards page to use the widget instead of the temporary delete button

Ali-Alkhars committed:
Handle categoryItem delete exceptions

Ali-Alkhars committed:
Create edit item button and page

- Refactor the admin choice boards page to use the new edit button widget
- Full implementation of the edit button is not done yet

Ali-Alkhars committed:
Implement item edit depending on what user changed

- Add comments in edit button and delete button widgets files

Ali-Alkhars committed:
Refactor firebase and image picker functions into separate files

- Create a folder for helper classes
- Create a class to hold all firebase manipulation functions
- Create a class to hold image picker functions
- Add missing comments to the new files
- Refactor pages to use the helper classes

Ali-Alkhars committed:
Fix multipath update for editing categoryItems

- Edit hardcoded items to test the functionality

Ali-Alkhars committed:
merge master into branch

Ali-Alkhars committed:
Make hardcoded categories be supplied dynamically

- Move hardcoded data into a separate file
- Add hardcoded data for testing
- Edit the keys of all the buttons in the admin choice boards
- Fix the widget tests for the admin choice boards
- Extend the tests for the admin side menu

Ali-Alkhars committed:
Extend widget test for admin choice boards page to test navigation to new pages

- Edit some keys
- Correct some old tests

Ali-Alkhars committed:
Write unit tests for firebase_functions helper class

- Rename firestore_functions into firebase_functions
- Write the tests using Mockito and build_runner to mock database calls

Ali-Alkhars committed:
Write firebase_firestore unit tests

- Refactor firebase_firestore to take the objects of Firebase as a parameter to allow mocking for tests
- Minor UI edit to heros (text is centered)
- Added/edited keys to widgets
- Added libraries which mock Firebase for testing
- Start widget testing for add/edit choice board item widgets

Ali-Alkhars committed:
Write widget tests for add, edit and delete item buttons

- Refactor most classes to have the ability to take mock firebase functions
- Adjust previous tests to accommodate refactoring changes
- Move MockFirebaseAuthentication class into a separate file
- Minor changes to hardcoded data

Ali-Alkhars committed:
Reuse ImageSquare widget to show images instead of card

Ali-Alkhars committed:
adjust drag icon position

AymenBerb commented on lib/helpers/mock_firebase_authentication.dart:
```diff
@@ -0,0 +1,4 @@
+import 'package:firebase_auth/firebase_auth.dart';
+import 'package:mockito/mockito.dart';
+
+class MockFirebaseAuthentication extends Mock implements FirebaseAuth {}
```
What is this file for ?

Ali-Alkhars replied:
I need it to mock FirebaseAuthentication

absirgue commented on lib/widgets/hero_dialog_route.dart:
```diff
@@ -0,0 +1,46 @@
+/// taken from: https://youtu.be/Bxs8Zy2O4wk
+
+class HeroDialogRoute<T> extends PageRoute<T> {
+  /// {@macro hero_dialog_route}
```
delete this comment

absirgue reviewed (approved):
Awesome work sir, clean and very well tested.

Just this tiny comment to delete and it's perfect.

Ali-Alkhars committed:
remove unnecessary comment

Ali-Alkhars committed:
Merge branch 'master' into T-09-Item-Backend

- Edit tests to work with refactoring changes

Ali-Alkhars committed:
Adjust UI to theme changes

Digital-Brendan reviewed (approved):
Everything looks great - nicely designed and commented.

If I was being picky, I would give two particular functions more concise names, specifically `handleSavingItemToFirestore` and `handleEditingItemInFirestore`.

How about `saveItemToFirestore` or just `saveItem`?

Generally though, great code!

Ali-Alkhars committed:
edit method names

AymenBerb reviewed (changes requested):
<img width="1512" alt="Screenshot 2023-03-03 at 16 03 14" src="https://user-images.githubusercontent.com/99401670/222768503-868acb31-3939-44aa-ab62-ab089f8da281.png">

Can you make item's name as a more recgnizable text field ? and maybe reduce the size of the pop up. (it doesnt need to take the whole screen (maybe add paading left and right))

Ali-Alkhars committed:
Edit Hero dialog UI

- Increase the size of text labels in hero dialog
- Decrease the size of hero dialog from the sides
- Fix a bug related to logging out

AymenBerb reviewed (approved):
well done !

Ali-Alkhars committed:
Merge master and resolve conflicts

Ali-Alkhars committed:
Add loading indicator for adding/editing/deleting items

- Refactor delete item button method into a separate method
- Adjust delete item test to accommodate the changes
- Remove unused imports in edit_account.dart
